### DISCUSSION TITLE
Picker组件增强的一些想法，便于columns数据是异步获取的场景

### DISCUSSION THREAD

**Discussion by damonyoungcc** [ant-design/ant-design-mobile#5399] [Ideas]
## 现在的问题：
源于这个`issue`: [[Picker]组件增加loading属性](https://github.com/ant-design/ant-design-mobile/issues/4945)，目前大部分用户异步获取columns数据的时机应该是两个：
1. 在页面挂载时就发起异步请求然后更新state，点击按钮展示`Picker`，传入`columns`
2. 点击按钮时，触发异步请求，成功后展示`Picker`

我想用户想要`loading`属性，大致是想在点击按钮后触发接口请求，而不是在一开始页面`mount`时就加载过多数据。

## 解决方案讨论
之前对于`antd` 的`Select`做过一些二次封装，利用了`Select`组件的两个`API`: `notFoundContent` 和 `onDropdownVisibleChange`，效果[点这里](https://yangdepp.github.io/ant-design-plus/components/ant-design/async-select#%E5%9F%BA%E7%A1%80%E7%94%A8%E6%B3%95) 
思路就是：
1. 用户在点击`Select`，先展示下拉框
2. `onDropdownVisibleChange`监听到`open`为`true`时发起异步请求，此时`loading`状态置为为`true`
3. 然后给`notFoundContent`传一个加载中的样式占位，等异步请求结束，则展示正常的内容。

综上，是否可以给`Picker`组件增加`notFoundContent` 和 `onPopupVisibleChange`属性，这样用户在点击`Picker`后，可以正常弹出`Picker`，异步数据请求成功前，展示`columns`的地方可以展示用户传入的骨架屏或loading组件，这样体验可能会更好。

**Comment by awmleer** [comment]
`onPopupVisibleChange` 感觉可以换成 `onOpen`，对应着已有的 `onClose`

`notFoundContent` 可以叫 `placeholder`，显示的时机是如果 `columns` 为 `[]` 空数组？

**Reply by xiaoyao96** [reply]
> notFoundContent 可以叫 placeholder，显示的时机是如果 columns 为 [] 空数组？

如果接口返回的结果也是空数组，就会一直展示 `placeholder` ，这个似乎不太适用于用来传 loading 效果。

`placeholder` 语义应该更偏向于，当picker未选择任何值时，提示用户请选择。

所以总结下来两个问题：
1、`placeholder` 命名问题
2、如何解决接口返回也是空数组的问题

可以详细讨论下...

**Comment by awmleer** [comment]
有一种边界情况需要考虑，如果用户初次点开 Picker，此时 `columns` 为 `[]`，显示 `placeholder` 占位提示，这时用户立刻点击确定按钮，要触发 `onChange` 吗？如果触发的话，`onChange` 里的值应该是多少？

**Reply by xiaohuoni** [reply]
默认值？

**Reply by awmleer** [reply]
或者也可以在加载过程中禁用掉确定按钮，避免用户点击

**Reply by damonyoungcc** [reply]
`notFoundContent`相当于就是空数组的一个占位，就像现在点开`Picker`是一个空数组时不展示空白，展示一个没有数据的占位

> 加载过程中禁用掉确定按钮，避免用户点击

我以上说的这些都是像`Select`一样只提供给用户可二次封装的可能性，具体这种异步逻辑`Select`是不提供的，`Picker`我觉得可以参考一下，这个禁用掉确定按钮可以只提供功能，具体控制交给用户，或者官网提供一个例子。

**Comment by damonyoungcc** [comment]
基于上面我写了一点用法的伪代码

```tsx
function BasicDemo() {
  const [ visible, setVisible ] = useState(false);
  const [ loading, setLoading ] = useState(false);
  const [columns, setColumns] = useState([]);

  const onOpen = async () => {
    // 1、开启，2、columns长度为0，3、loading为false时请求
    if(!columns.length && !loading) {
      setLoading(true)
      const data = await mockdata();
      setColumns(data)
      setLoading(false)
    }
  }

  return (
    <>
      <Button
        onClick={() => {
          setVisible(true)
        }}
      >
        选择
      </Button>
      <Picker
        confirmDisabled={!loading}
        placeholder={<div>loading...</div>} // 只有在传的情况下展示，不传保持原先样式不变
        onOpen={onOpen}
        columns={columns}
        visible={visible}
        onClose={() => {
          setVisible(false)
        }}
      />
    </>
  )
}
```

**Reply by awmleer** [reply]
`onOpen` 里再加个判断……？`columns.length > 0` 了是不是就不用再重新请求一次了

**Reply by damonyoungcc** [reply]
是的，我改一下

**Reply by awmleer** [reply]
`placeholder` 如果不传的话，默认就啥都不展示吧……？这样就和之前版本行为一致了，相当于 `placeholder` 是完全新增的特性

**Reply by damonyoungcc** [reply]
哦哦，是的，不传确实应该什么都不展示

**Reply by damonyoungcc** [reply]
`loading`为true时也不应该发起请求，避免用户点击后`loading`时间长，直接关闭掉`Picker`，下次再点开就会重复请求。

**Reply by awmleer** [reply]
> `loading`为true时也不应该发起请求，避免用户点击后`loading`时间长，直接关闭掉`Picker`，下次再点开就会重复请求。

这可以用 ahooks 的 `useLockFn` 来解决，就很简单了，都不需要再额外维护一个 `loading` 状态：

```jsx
function BasicDemo() {
  const [ visible, setVisible ] = useState(false);
  const [columns, setColumns] = useState([]);

  const lazyLoadColumns = useLockFn(async () => {
    if(!columns.length) {
      setColumns(await mockdata())
    }
  })

  return (
    <>
      <Button
        onClick={() => {
          setVisible(true)
        }}
      >
        选择
      </Button>
      <Picker
        placeholder={<div>loading...</div>} // 只有在传的情况下展示，不传保持原先样式不变
        onOpen={lazyLoadColumns}
        columns={columns}
        visible={visible}
        onClose={() => {
          setVisible(false)
        }}
      />
    </>
  )
}
```

**Reply by damonyoungcc** [reply]
嗯嗯，我试着改一下，这个pr我来试试吧

**Reply by damonyoungcc** [reply]
`confirmDisabled`这个我觉得也不用提供，原因有两点：
1. 源码中，`confirmText`是由`a`标签包裹，上面加`disabled`属性TS会报错，更改标签会造成`breakchange`，万一用户魔改了这个a标签的样式
2. 当前的`Picker`其实已经存在这样的场景： 用户在页面挂载时发起异步请求，请求时间很长，这时候用户点开了`Picker`直接点确定。所以这个我感觉不提供也行。

**Reply by damonyoungcc** [reply]
`onOpen`也不需要，因为`Popup`已经有了`afterShow`😄

**Reply by awmleer** [reply]
`afterShow` 触发的时机会晚一些（动画完全结束之后），还是得需要`onShow` 或者 `onOpen` 的（如果考虑到和已有的 afterShow 命名一致，那叫 onShow 比较合适一些了）
